File: DynamoDBObjectMapper-Sample/Objective-C/DynamoDBSample/DDBDynamoDBManager.m
/*
 * Copyright 2010-2015 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *  http://aws.amazon.com/apache2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

#import "DDBDynamoDBManager.h"
#import <AWSDynamoDB/AWSDynamoDB.h>
#import "Constants.h"

@implementation DDBDynamoDBManager

+ (AWSTask *)describeTable {
    AWSDynamoDB *dynamoDB = [AWSDynamoDB defaultDynamoDB];

    // See if the test table exists.
    AWSDynamoDBDescribeTableInput *describeTableInput = [AWSDynamoDBDescribeTableInput new];
    describeTableInput.tableName = AWSSampleDynamoDBTableName;
    return [dynamoDB describeTable:describeTableInput];
}

+ (AWSTask *)createTable {
    AWSDynamoDB *dynamoDB = [AWSDynamoDB defaultDynamoDB];

    // Create the test table.
    AWSDynamoDBAttributeDefinition *hashKeyAttributeDefinition = [AWSDynamoDBAttributeDefinition new];
    hashKeyAttributeDefinition.attributeName = @"UserId";
    hashKeyAttributeDefinition.attributeType = AWSDynamoDBScalarAttributeTypeS;

    AWSDynamoDBKeySchemaElement *hashKeySchemaElement = [AWSDynamoDBKeySchemaElement new];
    hashKeySchemaElement.attributeName = @"UserId";
    hashKeySchemaElement.keyType = AWSDynamoDBKeyTypeHash;

    AWSDynamoDBAttributeDefinition *rangeKeyAttributeDefinition = [AWSDynamoDBAttributeDefinition new];
    rangeKeyAttributeDefinition.attributeName = @"GameTitle";
    rangeKeyAttributeDefinition.attributeType = AWSDynamoDBScalarAttributeTypeS;

    AWSDynamoDBKeySchemaElement *rangeKeySchemaElement = [AWSDynamoDBKeySchemaElement new];
    rangeKeySchemaElement.attributeName = @"GameTitle";
    rangeKeySchemaElement.keyType = AWSDynamoDBKeyTypeRange;

    //Add non-key attributes
    AWSDynamoDBAttributeDefinition *topScoreAttrDef = [AWSDynamoDBAttributeDefinition new];
    topScoreAttrDef.attributeName = @"TopScore";
    topScoreAttrDef.attributeType = AWSDynamoDBScalarAttributeTypeN;
    
    AWSDynamoDBAttributeDefinition *winsAttrDef = [AWSDynamoDBAttributeDefinition new];
    winsAttrDef.attributeName = @"Wins";
    winsAttrDef.attributeType = AWSDynamoDBScalarAttributeTypeN;
    
    AWSDynamoDBAttributeDefinition *lossesAttrDef = [AWSDynamoDBAttributeDefinition new];
    lossesAttrDef.attributeName = @"Losses";
    lossesAttrDef.attributeType = AWSDynamoDBScalarAttributeTypeN;

    AWSDynamoDBProvisionedThroughput *provisionedThroughput = [AWSDynamoDBProvisionedThroughput new];
    provisionedThroughput.readCapacityUnits = @5;
    provisionedThroughput.writeCapacityUnits = @5;

    //Create Global Secondary Index
    NSArray *rangeKeyArray = @[@"TopScore",@"Wins",@"Losses"];
    NSMutableArray *gsiArray = [NSMutableArray new];
    for (NSString *rangeKey in rangeKeyArray) {
        AWSDynamoDBGlobalSecondaryIndex *gsi = [AWSDynamoDBGlobalSecondaryIndex new];
        
        AWSDynamoDBKeySchemaElement *gsiHashKeySchema = [AWSDynamoDBKeySchemaElement new];
        gsiHashKeySchema.attributeName = @"GameTitle";
        gsiHashKeySchema.keyType = AWSDynamoDBKeyTypeHash;
        
        AWSDynamoDBKeySchemaElement *gsiRangeKeySchema = [AWSDynamoDBKeySchemaElement new];
        gsiRangeKeySchema.attributeName = rangeKey;
        gsiRangeKeySchema.keyType = AWSDynamoDBKeyTypeRange;
        
        AWSDynamoDBProjection *gsiProjection = [AWSDynamoDBProjection new];
        gsiProjection.projectionType = AWSDynamoDBProjectionTypeAll;
        
        gsi.keySchema = @[gsiHashKeySchema,gsiRangeKeySchema];
        gsi.indexName = rangeKey;
        gsi.projection = gsiProjection;
        gsi.provisionedThroughput = provisionedThroughput;

        [gsiArray addObject:gsi];
    }
    

    //Create TableInput
    AWSDynamoDBCreateTableInput *createTableInput = [AWSDynamoDBCreateTableInput new];
    createTableInput.tableName = AWSSampleDynamoDBTableName;
    createTableInput.attributeDefinitions = @[hashKeyAttributeDefinition, rangeKeyAttributeDefinition, topScoreAttrDef, winsAttrDef,lossesAttrDef];
    createTableInput.keySchema = @[hashKeySchemaElement, rangeKeySchemaElement];
    createTableInput.provisionedThroughput = provisionedThroughput;
    createTableInput.globalSecondaryIndexes = gsiArray;

    return [[dynamoDB createTable:createTableInput] continueWithSuccessBlock:^id(AWSTask *task) {
        if (task.result) {
            // Wait for up to 4 minutes until the table becomes ACTIVE.

            AWSDynamoDBDescribeTableInput *describeTableInput = [AWSDynamoDBDescribeTableInput new];
            describeTableInput.tableName = AWSSampleDynamoDBTableName;
            task = [dynamoDB describeTable:describeTableInput];

            for(int32_t i = 0; i < 16; i++) {
                task = [task continueWithSuccessBlock:^id(AWSTask *task) {
                    AWSDynamoDBDescribeTableOutput *describeTableOutput = task.result;
                    AWSDynamoDBTableStatus tableStatus = describeTableOutput.table.tableStatus;
                    if (tableStatus == AWSDynamoDBTableStatusActive) {
                        return task;
                    }

                    sleep(15);
                    return [dynamoDB describeTable:describeTableInput];
                }];
            }
        }
        
        return task;
    }];
}

@end

@implementation DDBTableRow

+ (NSString *)dynamoDBTableName {
    return AWSSampleDynamoDBTableName;
}

+ (NSString *)hashKeyAttribute {
    return @"UserId";
}

+ (NSString *)rangeKeyAttribute {
    return @"GameTitle";
}

@end

@implementation DDBTableRowTopScore

+ (NSString *)dynamoDBTableName {
    return AWSSampleDynamoDBTableName;
}

+ (NSString *)hashKeyAttribute {
    return @"GameTitle";
}

+ (NSString *)rangeKeyAttribute {
    return @"TopScore";
}

@end

@implementation DDBTableRowWins

+ (NSString *)dynamoDBTableName {
    return AWSSampleDynamoDBTableName;
}

+ (NSString *)hashKeyAttribute {
    return @"GameTitle";
}

+ (NSString *)rangeKeyAttribute {
    return @"Wins";
}

@end

@implementation DDBTableRowLosses

+ (NSString *)dynamoDBTableName {
    return AWSSampleDynamoDBTableName;
}

+ (NSString *)hashKeyAttribute {
    return @"GameTitle";
}

+ (NSString *)rangeKeyAttribute {
    return @"Losses";
}

@end

File: DynamoDBObjectMapper-Sample/Swift/DynamoDBSampleSwift/DDBDynamoDBManager.swift
/*
 * Copyright 2010-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *  http://aws.amazon.com/apache2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import Foundation
import AWSDynamoDB

let AWSSampleDynamoDBTableName = "DynamoDB-OM-SwiftSample"

class DDBDynamoDBManger : NSObject {
    class func describeTable() -> AWSTask<AnyObject> {
        let dynamoDB = AWSDynamoDB.default()

        // See if the test table exists.
        let describeTableInput = AWSDynamoDBDescribeTableInput()
        describeTableInput?.tableName = AWSSampleDynamoDBTableName
        return dynamoDB.describeTable(describeTableInput!) as! AWSTask<AnyObject>
    }

    class func createTable() -> AWSTask<AnyObject> {
        let dynamoDB = AWSDynamoDB.default()

        //Create the test table
        let hashKeyAttributeDefinition = AWSDynamoDBAttributeDefinition()
        hashKeyAttributeDefinition?.attributeName = "UserId"
        hashKeyAttributeDefinition?.attributeType = AWSDynamoDBScalarAttributeType.S

        let hashKeySchemaElement = AWSDynamoDBKeySchemaElement()
        hashKeySchemaElement?.attributeName = "UserId"
        hashKeySchemaElement?.keyType = AWSDynamoDBKeyType.hash

        let rangeKeyAttributeDefinition = AWSDynamoDBAttributeDefinition()
        rangeKeyAttributeDefinition?.attributeName = "GameTitle"
        rangeKeyAttributeDefinition?.attributeType = AWSDynamoDBScalarAttributeType.S

        let rangeKeySchemaElement = AWSDynamoDBKeySchemaElement()
        rangeKeySchemaElement?.attributeName = "GameTitle"
        rangeKeySchemaElement?.keyType = AWSDynamoDBKeyType.range

        //Add non-key attributes
        let topScoreAttrDef = AWSDynamoDBAttributeDefinition()
        topScoreAttrDef?.attributeName = "TopScore"
        topScoreAttrDef?.attributeType = AWSDynamoDBScalarAttributeType.N

        let winsAttrDef = AWSDynamoDBAttributeDefinition()
        winsAttrDef?.attributeName = "Wins"
        winsAttrDef?.attributeType = AWSDynamoDBScalarAttributeType.N

        let lossesAttrDef = AWSDynamoDBAttributeDefinition()
        lossesAttrDef?.attributeName = "Losses"
        lossesAttrDef?.attributeType = AWSDynamoDBScalarAttributeType.N

        let provisionedThroughput = AWSDynamoDBProvisionedThroughput()
        provisionedThroughput?.readCapacityUnits = 5
        provisionedThroughput?.writeCapacityUnits = 5

        //Create Global Secondary Index
        let rangeKeyArray = ["TopScore","Wins","Losses"]
        var gsiArray = [AWSDynamoDBGlobalSecondaryIndex]()

        for rangeKey in rangeKeyArray {
            let gsi = AWSDynamoDBGlobalSecondaryIndex()

            let gsiHashKeySchema = AWSDynamoDBKeySchemaElement()
            gsiHashKeySchema?.attributeName = "GameTitle"
            gsiHashKeySchema?.keyType = AWSDynamoDBKeyType.hash

            let gsiRangeKeySchema = AWSDynamoDBKeySchemaElement()
            gsiRangeKeySchema?.attributeName = rangeKey
            gsiRangeKeySchema?.keyType = AWSDynamoDBKeyType.range

            let gsiProjection = AWSDynamoDBProjection()
            gsiProjection?.projectionType = AWSDynamoDBProjectionType.all;

            gsi?.keySchema = [gsiHashKeySchema!,gsiRangeKeySchema!];
            gsi?.indexName = rangeKey;
            gsi?.projection = gsiProjection;
            gsi?.provisionedThroughput = provisionedThroughput;

            gsiArray.append(gsi!)
        }

        //Create TableInput
        let createTableInput = AWSDynamoDBCreateTableInput()
        createTableInput?.tableName = AWSSampleDynamoDBTableName;
        createTableInput?.attributeDefinitions = [hashKeyAttributeDefinition!, rangeKeyAttributeDefinition!, topScoreAttrDef!, winsAttrDef!,lossesAttrDef!]
        createTableInput?.keySchema = [hashKeySchemaElement!, rangeKeySchemaElement!]
        createTableInput?.provisionedThroughput = provisionedThroughput
        createTableInput?.globalSecondaryIndexes = gsiArray

        return dynamoDB.createTable(createTableInput!).continueOnSuccessWith(block: { task -> AnyObject? in
            guard task.result != nil else {
                return task
            }

            // Wait for up to 4 minutes until the table becomes ACTIVE.

            let describeTableInput = AWSDynamoDBDescribeTableInput()
            describeTableInput?.tableName = AWSSampleDynamoDBTableName;
            let describeTask = dynamoDB.describeTable(describeTableInput!)

            var localTask:AWSTask<AnyObject>?
            for _ in 0...15 {
                localTask = describeTask.continueOnSuccessWith(block: { task -> AnyObject? in
                    let describeTableOutput:AWSDynamoDBDescribeTableOutput = task.result!
                    let tableStatus = describeTableOutput.table!.tableStatus
                    if tableStatus == AWSDynamoDBTableStatus.active {
                        return task
                    }

                    sleep(15)
                    return dynamoDB .describeTable(describeTableInput!)
                })
            }

            return localTask
        })

    }
}

class DDBTableRow :AWSDynamoDBObjectModel ,AWSDynamoDBModeling  {

    var UserId:String?
    var GameTitle:String?

    //set the default values of scores, wins and losses to 0
    var TopScore:NSNumber? = 0
    var Wins:NSNumber? = 0
    var Losses:NSNumber? = 0

    //should be ignored according to ignoreAttributes
    var internalName:String?
    var internalState:NSNumber?

    class func dynamoDBTableName() -> String {
        return AWSSampleDynamoDBTableName
    }

    class func hashKeyAttribute() -> String {
        return "UserId"
    }
    
    class func rangeKeyAttribute() -> String {
        return "GameTitle"
    }
    
    class func ignoreAttributes() -> [String] {
        return ["internalName", "internalState"]
    }
}

